smithy-swift-codegen/src/main/kotlin/software/amazon/smithy/swift/codegen/waiters/JMESPathVisitor.kt

```diff
@@ -0,0 +1,412 @@
+            RuntimeType.NUMBER -> return addTempVar(Variable("number", false, doubleShape), "Double(\$L)", expression.expectNumberValue())
+            RuntimeType.BOOLEAN -> return addTempVar(Variable("bool", false, boolShape), "\$L", expression.expectBooleanValue())
+            RuntimeType.NULL -> return Variable("nil", true, boolShape)
+            else -> throw Exception("Expression type $expression is unsupported")
+        }
+    }
+
+    override fun visitMultiSelectHash(expression: MultiSelectHashExpression): Variable {
+        throw Exception("MultiSelectHashExpression is unsupported")
+    }
+
+    // Render a JMESPath multi-select to an array.
+    // All expressions must result in the same type or a Swift compile error will occur.
+    override fun visitMultiSelectList(expression: MultiSelectListExpression): Variable {
+        val listName = uniqueTempVarName("multiSelect")
+        var innerShape = currentExpression.shape
+        val multiSelectVars = expression.expressions.map { inner ->
+            val result = inner.accept(this)
+            innerShape = result.shape
+            return result
+        }
+        writer.openBlock("let \$L = [", "]", listName) {
+            writer.write(multiSelectVars.joinToString { ",\n" })
+        }
+        val memberShape = MemberShape.builder().id("smithy.swift.synthetic#MultiSelectListElement").build()
+        val multiSelectListShape = ListShape.builder()
+            .id("smithy.swift.synthetic#MultiSelectList")
+            .member(memberShape)
+            .build()
+        return Variable(listName, false, multiSelectListShape)
+    }
+
+    // Negates the passed expression.
+    // The passed expression must be Boolean, else a Swift compile error will occur.
+    override fun visitNot(expression: NotExpression): Variable {
+        val expressionToNegate = expression.expression!!.accept(this)
+        val negatedVar = Variable("negated", false, expressionToNegate.shape)
+        return addTempVar(negatedVar, "!\$L", expressionToNegate.name)
+    }
+
+    // Converts a JSON object / Swift dictionary into an array of its values.
+    override fun visitObjectProjection(expression: ObjectProjectionExpression): Variable {
+        val original = expression.left!!.accept(this)
+        return when (original.shape) {
+            is MapShape -> {
+                val valueShape = model.expectShape(original.shape.value.target)
+                val projectionShape = ListShape.builder()
+                    .id("smithy.swift.synthetic#ObjectProjection")
+                    .member(valueShape.toShapeId())
+                    .build()
+                var projectionVar = Variable("projection", original.isOptional, projectionShape)
+                if (original.isOptional) {
+                    projectionVar = addTempVar(projectionVar, "\$L.flatMap { Array($$0.values) }", original.name)
```

Comment:
why `flatMap` here? should this just be `map` instead?

Reply:
This is the optional version of `flatMap`, not the array version.

Reply:
https://developer.apple.com/documentation/swift/optional/flatmap(_:)#

Reply:
Right, so shouldn't we use `map` here since the closure returns a non-optional?

Reply:
The closure returns a non-optional, but the return value of `flatMap` is optional.

Reply:
Changed to `map`